### PR TITLE
[16.0][OU-ADD] analytic: Migration done

### PR DESCRIPTION
**Pull request by daiduongnguyen-odoo**
Need test with account also

### CODE REVIEW

File: openupgrade_scripts/scripts/analytic/16.0.1.1/upgrade_analysis_work.txt [hunk 1]
```diff
@@ -0,0 +1,94 @@
+---Models in module 'analytic'---
+obsolete model account.analytic.distribution
+obsolete model account.analytic.tag
+new model account.analytic.applicability
+new model account.analytic.distribution.model
+new model analytic.mixin [abstract]
+# NOTHING TO DO
```
Comment (daiduongnguyen-odoo) [inline openupgrade_scripts/scripts/analytic/16.0.1.1/upgrade_analysis_work.txt:7]: This analysis work file is base on newly generation at PR : https://github.com/OCA/OpenUpgrade/pull/4042. I did not edit the original `upgrade_analysis` only `ugrade_analysis_work`

File: openupgrade_scripts/scripts/analytic/16.0.1.1/pre-migration.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+]
+_models_renames = [("account.analytic.group", "account.analytic.plan")]
+_tables_renames = [("account_analytic_group", "account_analytic_plan")]
+
+
+def analytic_account_set_group_id_if_null(env):
+    """
+    We would like to fill for group_id column before start renaming it into plan_id
+    """
+    openupgrade.logged_query(
```
Comment (daiduongnguyen-odoo) [inline openupgrade_scripts/scripts/analytic/16.0.1.1/pre-migration.py]: I think in production database it is quite rare to have `analytic_account` with empty `analytic_group` right? but still IMO, it need to be filled before changing it into `plan`

File: openupgrade_scripts/scripts/analytic/16.0.1.1/pre-migration.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        WHERE aaa.id IN (SELECT id FROM tmp);
+        """,
+    )
+
+
+def create_root_plan_and_prefill_value_for_analytic_plan(env):
+    """
+    Pre fill value for column root_plan_id using the logic in its compute method to avoid computed by ORM
+    """
+    openupgrade.logged_query(
```
Comment (daiduongnguyen-odoo) [inline openupgrade_scripts/scripts/analytic/16.0.1.1/pre-migration.py]: To avoid computing using ORM for performance

File: openupgrade_scripts/scripts/analytic/16.0.1.1/upgrade_analysis_work.txt [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+analytic     / account.analytic.account / group_id (many2one)           : DEL relation: account.analytic.group
+analytic     / account.analytic.account / plan_id (many2one)            : NEW relation: account.analytic.plan, required
+# DONE pre-migration: rename field and fill value for null row
+
+analytic     / account.analytic.account / root_plan_id (many2one)       : NEW relation: account.analytic.plan, isfunction: function, stored
+# DONE pre-migration: create column and pre-fill value to avoid computed by ORM
+
+analytic     / account.analytic.applicability / analytic_plan_id (many2one)   : NEW relation: account.analytic.plan
+analytic     / account.analytic.applicability / applicability (selection)     : NEW required, selection_keys: ['mandatory', 'optional', 'unavailable']
+analytic     / account.analytic.applicability / business_domain (selection)   : NEW required, selection_keys: ['general']
```
Comment (daiduongnguyen-odoo) [inline openupgrade_scripts/scripts/analytic/16.0.1.1/upgrade_analysis_work.txt:26]: I think this one and the one i comment below should be handled in `account` module

File: openupgrade_scripts/scripts/analytic/16.0.1.1/upgrade_analysis_work.txt [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+analytic     / account.analytic.distribution / percentage (float)            : DEL required
+analytic     / account.analytic.distribution / tag_id (many2one)             : DEL relation: account.analytic.tag, required
+# NOTHING TO DO
+
+analytic     / account.analytic.distribution.model / analytic_distribution (json)  : NEW hasdefault: compute
+analytic     / account.analytic.distribution.model / analytic_distribution_search (json): NEW
+analytic     / account.analytic.distribution.model / analytic_precision (integer)  : NEW hasdefault: default
+analytic     / account.analytic.distribution.model / company_id (many2one)         : NEW relation: res.company, hasdefault: default
+analytic     / account.analytic.distribution.model / partner_category_id (many2one): NEW relation: res.partner.category
+analytic     / account.analytic.distribution.model / partner_id (many2one)         : NEW relation: res.partner
```
Comment (daiduongnguyen-odoo) [inline openupgrade_scripts/scripts/analytic/16.0.1.1/upgrade_analysis_work.txt:39]: Same above should handle in account module

**Comment by daiduongnguyen-odoo** [comment]
Anybody help me review pls :((

**Comment by legalsylvain** [comment]
@remytms, could you take a look over there ? thanks 
@duong77476 thanks for your contribution !

**Comment by daiduongnguyen-odoo** [comment]
\ocabot rebase

**Review by ngochung207 — APPROVED** [review]

File: openupgrade_scripts/apriori.py [hunk 6]
```diff
@@ -21,6 +21,21 @@
...
+    "to_stock_backdate": "viin_stock_backdate",
+    "to_stock_picking_backdate": "viin_stock_account_backdate",
+    "to_uom_mail_thread": "viin_mail_thread_uom",
+    "viin_google_drive_support_oauth2": "viin_google_drive",
+    "viin_product_categ_mail_thread_purchase": "viin_mail_thread_purchase",
+    "viin_product_categ_mail_thread_stock_account": "viin_mail_thread_stock_account",
+    "viin_webp": "viin_web_editor",
+    "viin_user_assignment_log_project": "test_viin_user_assignment_log_project",
+    # Viindoo/enterprise
+    # Viindoo/odoo-tvtma
```
Comment (katyukha) [inline openupgrade_scripts/apriori.py]: Do we need this in OCA?

File: openupgrade_scripts/apriori.py [hunk 7]
```diff
@@ -55,6 +70,24 @@
...
+    "viin_hr_holidays_accrual_plan": "viin_hr_holidays",
+    "viin_hr_employee_resource_calendar": "viin_hr_contract",
+    "viin_project_kanban_state_notification": "viin_project",
+    "viin_project_update": "viin_project",
+    "viin_website_image_optimization_disable": "viin_web_editor",
+    # Viindoo/enterprise
+    "viin_website_seo_advisor_patch1": "viin_website_seo_advisor",
+    "viin_website_patch1": "viin_website",
+    "viin_website_helpdesk_patch1": "viin_website_helpdesk",
+    # Viindoo/odoo-tvtma
```
Comment (katyukha) [inline openupgrade_scripts/apriori.py]: Do we need this in OCA OpenUpgrade?

File: openupgrade_scripts/apriori.py [hunk 8]
```diff
@@ -76,11 +109,24 @@
...
+    # Viindoo/tvtmaaddons
+    "approval.request": "viin.approval.request",
+    "helpdesk.sla": "viin.helpdesk.sla",
+    "helpdesk.stage": "viin.helpdesk.stage",
+    "helpdesk.tag": "viin.helpdesk.tag",
+    "helpdesk.team": "viin.helpdesk.team",
+    "helpdesk.ticket": "viin.helpdesk.ticket",
+    "helpdesk.ticket.type": "viin.helpdesk.ticket.type",
+    # Viindoo/enterprise
+    # Viindoo/odoo-tvtma
```
Comment (katyukha) [inline openupgrade_scripts/apriori.py]: Do we need this in OCA?

File: openupgrade_scripts/apriori.py [hunk 9]
```diff
@@ -76,11 +109,24 @@
...
 }
 
 # only used here for upgrade_analysis
 merged_models = {
     # odoo
     "gift.card": "loyalty.card",
     # OCA/...
+    # Viindoo/tvtmaaddons
+    # Viindoo/enterprise
+    # Viindoo/odoo-tvtma
```
Comment (katyukha) [inline openupgrade_scripts/apriori.py]: Do we need this in OCA?

File: openupgrade_framework/odoo_patch/odoo/modules/loading.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            msg_all = msg("Table name", "Old row count", "New row count")
+            for r in result:
+                msg_all += msg(r["table_name"], r["old_row_count"], r["new_row_count"])
+            _logger.warning(
+                "The following tables have different number of rows:\n" "%s\n" % msg_all
+            )
+
+
+load_modules._original_method = odoo.modules.load_modules
+odoo.modules.load_modules = load_modules
```
Comment (katyukha) [inline openupgrade_framework/odoo_patch/odoo/modules/loading.py]: Why this function redefined?
May be it have sense to make it as separate PR if needed?
Also, i think comment or docstring have to be added that describes purpose of this function redefinition.

File: .github/workflows/test.yml [hunk 11]
```diff
@@ -88,7 +88,7 @@ jobs:
           sudo npm install -g less less-plugin-clean-css
           pip install -q -r odoo/requirements.txt
           pip install --ignore-installed \
-              git+https://github.com/OCA/openupgradelib.git@master
+              git+https://github.com/duong77476/openupgradelib.git@master_imp_boostrap4_to_5_transformation
```
Comment (katyukha) [inline .github/workflows/test.yml]: Why changing it here?

**Review by katyukha — COMMENTED** [review]
It seems that some chanes not related to OCA were added to this PR.

File: openupgrade_scripts/apriori.py (hunk 6 above)
Comment (jcdrubay) [reply openupgrade_scripts/apriori.py]: Related discussion: https://github.com/OCA/OpenUpgrade/discussions/3245

Comment (pedrobaeza) [reply openupgrade_scripts/apriori.py]: Non OCA modules should be added in a fork, and use such fork in your projects, but they shouldn't land on main OpenUpgrade repository. You can use git-aggregator for merging main repo with these changes transparently.

Example of custom migration scripts (in this case, migrating from purchased modules to OCA ones from one bounced client): https://github.com/Tecnativa/OpenUpgrade/commit/ac0334610b325aa81e4c1c5d89daa0e608e2b524

**Review by remi-filament — APPROVED** [review]
Thanks @duong77476, tested OK on our database (where typically we did not use analytic groups :))

**Comment by pedrobaeza** [comment]
Thanks for the migration scripts, Nguyễn. Unfortunately, you haven't allowed to push changes on your branch for the maintainers, so I have superseded this PR at #4188, respecting your attribution, rebasing on top of current branch, and adding the following changes:

- Database structure changes should be always applied before business logic changes.
- No need to implement literally the compute root_plan_id, as we are creating it for the first time. We just assign it on the accounts.
- Restructured dummy plan creation, for having the ID available for several queries and not having to make a fuzzy select.
- For a smooth migration, the default_applicability should be optional, so that users can still continue applying their accounts into documents.
- Rename default plan to 'Legacy' for better understanding.
